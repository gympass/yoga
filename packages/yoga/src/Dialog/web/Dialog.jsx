import React, { useEffect, useCallback } from 'react';
import { createPortal } from 'react-dom';
import styled from 'styled-components';
import { func, bool, node, number, string } from 'prop-types';

import { usePortal, useCombinedRefs } from '../../hooks';
import { Card } from '../..';

export const StyledDialog = styled(Card)`
  ${({
    onClose,
    theme: {
      yoga: {
        components: { dialog },
      },
    },
  }) => `
  padding: ${onClose ? dialog.padding.withCloseButton : dialog.padding.top}px
  ${dialog.padding.default}px
  ${dialog.padding.default}px;

  width: ${dialog.width.default}px;
  min-height: ${dialog.height.min}px;
  border-radius: ${dialog.border.radius}px;
  `}
  display: flex;
  align-items: center;
  flex-direction: column;
`;

const Overlay = styled.div`
  ${({
    zIndex,
    theme: {
      yoga: {
        components: { dialog },
      },
    },
  }) => `
  display: flex;
  z-index: ${zIndex};
  justify-content: center;
  align-items: center;

  position: fixed;
  top: ${dialog.position.default};
  right: ${dialog.position.default};
  left: ${dialog.position.default};
  bottom: ${dialog.position.default};

  background-color: rgba(35, 27, 34, 0.48);
  `}
`;

const Dialog = React.forwardRef(
  ({ isOpen, children, dialogId, onClose, zIndex, ...props }, forwardedRef) => {
    const dialogRef = useCombinedRefs(forwardedRef);
    const dialogElement = usePortal(dialogId ?? 'dialog');

    const closeDialog = useCallback(
      e => {
        if (dialogRef.current === e.target && isOpen && onClose) {
          onClose(e);
        }
        return true;
      },
      [onClose, isOpen],
    );

    const keyPress = useCallback(
      e => {
        if (e.key === 'Escape' && isOpen && onClose) {
          onClose(e);
        }
        return true;
      },
      [onClose, isOpen],
    );

    useEffect(() => {
      document.addEventListener('keydown', keyPress);

      return () => document.removeEventListener('keydown', keyPress);
    }, [keyPress]);

    return isOpen ? (
      createPortal(
        <Overlay
          onClick={closeDialog}
          onClose={onClose}
          ref={dialogRef}
          zIndex={zIndex}
        >
          <StyledDialog onClose={onClose} {...props}>
            {children}
          </StyledDialog>
        </Overlay>,
        dialogElement,
      )
    ) : (
      <></>
    );
  },
);

Dialog.propTypes = {
  /** Custom id to allow creating isolated portal nodes */
  dialogId: string,
  /** Control the dialog visibility. */
  isOpen: bool,
  /** Function to close the dialog. */
  onClose: func,
  backHandler: func,
  zIndex: number,
  children: node.isRequired,
};

Dialog.defaultProps = {
  isOpen: false,
  onClose: undefined,
  backHandler: undefined,
  zIndex: 3,
  dialogId: undefined,
};

Dialog.displayName = 'Dialog';

export default Dialog;
