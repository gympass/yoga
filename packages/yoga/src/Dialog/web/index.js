import Dialog from './Dialog';
import Header from './Header';
import Content from './Content';
import Footer from './Footer';

const ExportDialog = Object.assign(Dialog, {
  Header,
  Content,
  Footer,
});

export default ExportDialog;
