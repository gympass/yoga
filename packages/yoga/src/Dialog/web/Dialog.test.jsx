import React from 'react';
import { screen, render, fireEvent, cleanup } from '@testing-library/react';

import { ThemeProvider, Button } from '../..';

import Dialog from '.';

describe('<Dialog />', () => {
  afterEach(cleanup);

  it('should match snapshot', () => {
    const { baseElement } = render(
      <ThemeProvider>
        <Dialog isOpen>
          <Dialog.Header>Title</Dialog.Header>
          <Dialog.Content>Subtitle</Dialog.Content>
          <Dialog.Footer>
            <Button secondary>Ok, got it</Button>
          </Dialog.Footer>
        </Dialog>
      </ThemeProvider>,
    );

    expect(baseElement).toMatchSnapshot();
  });

  it('should match snapshot with close button', () => {
    const { baseElement } = render(
      <ThemeProvider>
        <Dialog isOpen onClose={jest.fn()}>
          <Dialog.Header onClose={jest.fn()}>Title</Dialog.Header>
          <Dialog.Content>Subtitle</Dialog.Content>
          <Dialog.Footer>
            <Button secondary>Ok, got it</Button>
          </Dialog.Footer>
        </Dialog>
      </ThemeProvider>,
    );

    expect(baseElement).toMatchSnapshot();
  });

  it('should render a minimal dialog', () => {
    const onActionMock = jest.fn();

    render(
      <ThemeProvider>
        <Dialog isOpen>
          <Dialog.Header>Title</Dialog.Header>
          <Dialog.Content>Subtitle</Dialog.Content>
          <Dialog.Footer>
            <Button onClick={onActionMock} secondary>
              Ok, got it
            </Button>
          </Dialog.Footer>
        </Dialog>
      </ThemeProvider>,
    );

    screen.getByText('Title');
    screen.getByText('Subtitle');

    const button = screen.getByRole('button', { name: /Ok, got it/i });

    fireEvent.click(button);

    expect(onActionMock).toHaveBeenCalledTimes(1);
  });

  it('should render with close button', () => {
    const onCloseMock = jest.fn();

    render(
      <ThemeProvider>
        <Dialog isOpen onClose={onCloseMock}>
          <Dialog.Header onClose={onCloseMock}>Title</Dialog.Header>
        </Dialog>
      </ThemeProvider>,
    );
    const button = screen.getByRole('button');

    fireEvent.click(button);

    expect(onCloseMock).toHaveBeenCalledTimes(1);
  });

  it('should hide the close button when hideCloseButton prop is true', () => {
    render(
      <ThemeProvider>
        <Dialog isOpen hideCloseButton onClose={() => {}}>
          <Dialog.Header>Title</Dialog.Header>
        </Dialog>
      </ThemeProvider>,
    );
    expect(screen.queryByRole('button')).toBeNull();
  });

  it('should render content isolated by dialog id', () => {
    render(
      <ThemeProvider>
        <Dialog isOpen>
          <Dialog.Header>Title</Dialog.Header>
        </Dialog>

        <Dialog isOpen dialogId="custom-id">
          <Dialog.Header>Second Title</Dialog.Header>
        </Dialog>
      </ThemeProvider>,
    );

    expect(screen.getByText('Title')).toBeInTheDocument();
    expect(screen.getByText('Second Title')).toBeInTheDocument();
  });
});
