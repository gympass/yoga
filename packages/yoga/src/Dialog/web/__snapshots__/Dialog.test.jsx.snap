// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dialog /> should match snapshot 1`] = `
.c6 {
  box-sizing: border-box;
  text-align: center;
  outline: none;
  -webkit-transition: all 0.2s;
  transition: all 0.2s;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: auto;
  height: 48px;
  padding-left: 24px;
  padding-right: 24px;
  background-color: #231B22;
  border: none;
  border-radius: 9999px;
  color: #FFFFFF;
  font-size: 16px;
  font-weight: 500;
  font-family: Rubik;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
  line-height: 1;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c6 svg {
  width: 24px;
  height: 24px;
  fill: #FFFFFF;
  margin-right: 8px;
  -webkit-transition: fill 0.2s;
  transition: fill 0.2s;
}

.c6:not([disabled]):hover,
.c6:not([disabled]):focus {
  box-shadow: 0 4px 8px rgba(35,27,34,0.45);
}

.c6:active {
  background-color: rgba(35,27,34,0.75);
  color: #FFFFFF;
}

.c6:active svg {
  fill: #FFFFFF;
}

.c3 {
  margin-bottom: 24px;
  color: #231B22;
  font-size: 24px;
  font-weight: 500;
  text-align: center;
}

.c4 {
  margin-bottom: 32px;
  color: #6B6B78;
  font-size: 16px;
  font-weight: 400;
  text-align: center;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  gap: 24px;
}

.c1 {
  padding: 16px  16px  16px  16px;
  border-radius: 8px;
  background-color: #FFFFFF;
  box-shadow: 0 2px 4px -1px rgba(152,152,166,0.2),0 4px 5px 0px rgba(152,152,166,0.14),0 1px 10px 0px rgba(152,152,166,0.12);
}

.c2 {
  padding: 40px 32px 32px;
  width: 580px;
  min-height: 160px;
  border-radius: 16px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  z-index: 3;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: fixed;
  top: 0;
  right: 0;
  left: 0;
  bottom: 0;
  background-color: rgba(35,27,34,0.48);
}

<body>
  <div />
  <div
    id="dialog"
  >
    <div>
      <div
        class="c0"
      >
        <section
          class="c1 c2"
        >
          <header
            class="c3"
            color="text.primary"
          >
            Title
          </header>
          <div
            class="c4"
            color="text.secondary"
          >
            Subtitle
          </div>
          <footer
            class="c5"
            d="flex"
          >
            <button
              class="c6"
            >
              Ok, got it
            </button>
          </footer>
        </section>
      </div>
    </div>
  </div>
</body>
`;

exports[`<Dialog /> should match snapshot with close button 1`] = `
<body>
  <div />
  .c9 {
  box-sizing: border-box;
  text-align: center;
  outline: none;
  -webkit-transition: all 0.2s;
  transition: all 0.2s;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: auto;
  height: 48px;
  padding-left: 24px;
  padding-right: 24px;
  background-color: #231B22;
  border: none;
  border-radius: 9999px;
  color: #FFFFFF;
  font-size: 16px;
  font-weight: 500;
  font-family: Rubik;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
  line-height: 1;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c9 svg {
  width: 24px;
  height: 24px;
  fill: #FFFFFF;
  margin-right: 8px;
  -webkit-transition: fill 0.2s;
  transition: fill 0.2s;
}

.c9:not([disabled]):hover,
.c9:not([disabled]):focus {
  box-shadow: 0 4px 8px rgba(35,27,34,0.45);
}

.c9:active {
  background-color: rgba(35,27,34,0.75);
  color: #FFFFFF;
}

.c9:active svg {
  fill: #FFFFFF;
}

.c3 {
  margin-bottom: 24px;
  color: #231B22;
  font-size: 24px;
  font-weight: 500;
  text-align: center;
}

.c7 {
  margin-bottom: 32px;
  color: #6B6B78;
  font-size: 16px;
  font-weight: 400;
  text-align: center;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  gap: 24px;
}

.c4 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c6 {
  width: 24px;
  height: 24px;
}

.c5 {
  box-sizing: border-box;
  text-align: center;
  outline: none;
  -webkit-transition: all 0.2s;
  transition: all 0.2s;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: auto;
  height: 48px;
  padding-left: 24px;
  padding-right: 24px;
  background-color: #D8385E;
  border: none;
  border-radius: 9999px;
  color: #FFFFFF;
  font-size: 16px;
  font-weight: 500;
  font-family: Rubik;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
  line-height: 1;
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: #FFFFFF;
  color: #D8385E;
  padding: 0;
  width: 48px;
}

.c5 svg {
  width: 24px;
  height: 24px;
  fill: #FFFFFF;
  margin-right: 8px;
  -webkit-transition: fill 0.2s;
  transition: fill 0.2s;
}

.c5:not([disabled]):hover,
.c5:not([disabled]):focus {
  box-shadow: 0 4px 8px rgba(216,56,94,0.45);
}

.c5:active {
  background-color: rgba(216,56,94,0.75);
  color: #FFFFFF;
}

.c5:active svg {
  fill: #FFFFFF;
}

.c5 svg {
  fill: #D8385E;
}

.c5:active {
  background-color: rgba(255,255,255,0.75);
  color: rgba(216,56,94,0.75);
}

.c5:active svg {
  fill: rgba(216,56,94,0.75);
}

.c5:not([disabled]):hover,
.c5:not([disabled]):focus {
  box-shadow: 0 4px 8px rgba(255,255,255,0.45);
}

.c5 svg {
  width: unset;
  height: unset;
  margin-right: unset;
  -webkit-transition: fill 0.2s;
  transition: fill 0.2s;
}

.c1 {
  padding: 16px  16px  16px  16px;
  border-radius: 8px;
  background-color: #FFFFFF;
  box-shadow: 0 2px 4px -1px rgba(152,152,166,0.2),0 4px 5px 0px rgba(152,152,166,0.14),0 1px 10px 0px rgba(152,152,166,0.12);
}

.c2 {
  padding: 24px 32px 32px;
  width: 580px;
  min-height: 160px;
  border-radius: 16px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  z-index: 3;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: fixed;
  top: 0;
  right: 0;
  left: 0;
  bottom: 0;
  background-color: rgba(35,27,34,0.48);
}

<div
    id="dialog"
  >
    <div>
      <div
        class="c0"
      >
        <section
          class="c1 c2"
        >
          <header
            class="c3"
            color="text.primary"
          >
            <div
              class="c4"
              d="flex"
            >
              <button
                class="c5"
              >
                <svg
                  class="c6"
                  height="24"
                  width="24"
                />
              </button>
            </div>
            Title
          </header>
          <div
            class="c7"
            color="text.secondary"
          >
            Subtitle
          </div>
          <footer
            class="c8"
            d="flex"
          >
            <button
              class="c9"
            >
              Ok, got it
            </button>
          </footer>
        </section>
      </div>
    </div>
  </div>
</body>
`;
